h2o-core/src/main/java/water/util/FrameSizeMonitor.java

```diff
@@ -0,0 +1,151 @@
+package water.util;
+
+import water.*;
+import water.fvec.Frame;
+import water.fvec.NewChunk;
+import water.parser.FVecParseWriter;
+
+import java.util.HashMap;
+import java.util.Map;
+
+public class FrameSizeMonitor extends MrFun<FrameSizeMonitor> {
+
+    private static final String ENABLED_PROP = "util.frameSizeMonitor.enabled";
+    private static final String SAFE_COEF_PROP = "util.frameSizeMonitor.safetyCoefficient";
+    private static final String SAFE_FREE_MEM_DEFAULT_COEF = "0.1";
+
+    private static final boolean ENABLED;
+    private static final float SAFE_FREE_MEM_COEF;
+    private static final int SLEEP_MS = 100;
+    private static final int INTERVAL_MS = 1000;
+    
+    static {
+        ENABLED = H2O.getSysBoolProperty(ENABLED_PROP, true);
+        SAFE_FREE_MEM_COEF = Float.parseFloat(H2O.getSysProperty(SAFE_COEF_PROP, SAFE_FREE_MEM_DEFAULT_COEF));
+    }
+
+    private final Key<Job> _jobKey;
+    private final Map<FVecParseWriter, Long> writers = new HashMap<>();
+
+    public FrameSizeMonitor(Key<Job> jobKey) {
+        _jobKey = jobKey;
+    }
+
+    public static FrameSizeMonitor start(Key<Job> jobKey) {
+        Job<Frame> job = DKV.getGet(jobKey);
+        FrameSizeMonitor task = new FrameSizeMonitor(jobKey);
+        if (ENABLED) {
+            H2O.submitTask(new LocalMR(task, 1));
+        }
+        return task;
+    }
+
+    @Override
+    protected void map(int id) {
+        float nextProgress = 0.1f / 16f; // take 4 frequent measurements before stopping at every 10%
+        Job<Frame> job = DKV.getGet(_jobKey);
+        while (job.isRunning()) {
+            job = DKV.getGet(_jobKey); // refresh progress
+            float currentProgress = job.progress();
+            System.out.println("WAIT FOR NEXT " + nextProgress + " " + currentProgress);
+            if (currentProgress >= nextProgress) {
+                if (isMemoryUsageOverLimit(currentProgress, job)) {
+                    job.stop();
+                    break;
+                }
+                if (nextProgress < 0.1f) {
+                    nextProgress *= 2;
+                }
+                System.out.println("NEXT " + nextProgress);
+            }
+            try {
+                Thread.sleep(SLEEP_MS);
```

Comment:
Rarely, this might not be enough. The parsers tends to be really quick. 

We could do the checks before any memory is allocated directly in the thread touched. But that would require additional overhead, even if the monitoring feature would be disabled. Which is probably undesirable in high-perf core code. I guess hotspot could eliminate the performance impact once it kicks in.

Reply:
VERY rarely. Very. Just thinking out loud and pointing out the weaknesses I believe @honzasterba is well aware of.

Reply:
By the "affected thread", I mean the NewChunk itself - doing the test there before allocating new memory or doing memory changes (sparsity <-> density transitions etc.) would make sense ?

Reply:
One adjustment I can think of is to only use `SLEEP_MS` as a default value and adjust it based on the relative speed of the parser against free memory. This would be relatively cheap and easy check that could make this almost bulletproof. Not going over 100ms (current constant value, or any other), but in well-reasoned cases shorten the interval to be 100% sure.

Reply:
actually checking more frequently does not have much benefit because the progress does not change that much and also memory is usually allocated in big batches, to avoid false positives I have added a check for overall memory usage and only terminate when a certain overall memory usage is reached and it seems the predicted size will not fit